### PR TITLE
ci: fix previous release tag detection

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

The problem:
 - the `release.yml` created a PR https://github.com/cowprotocol/cowswap/pull/7554 which was merged
 - after that it created a similar PR with the same changelogs: https://github.com/cowprotocol/cowswap/pull/7555

The cause:
The converter (`commits-to-changesets.mjs`) uses the latest `release-*` tag as its baseline. That tag is only pushed at the end of the publish job, but the converter runs at the start of the same workflow. So the workflow run triggered by a release-PR merge sees an older `release-*` tag, walks back across commits the PR just consumed, and re-emits all of them as fresh changesets — spawning a phantom duplicate version PR. The publish job did eventually push release-20260521T072952Z at 07:29:52, but #7555 was already opened at 07:27:08.

The fix:
  - New baseline precedence: `BASELINE_REF` env → most recent `chore(main): release` commit reachable from HEAD (when newer than the last release tag) → latest `release-*` tag → none.
  - The `chore(main): release` merge commit is the canonical "everything up to here was just released" marker — available immediately on merge, no race with the publish job.

# To Test

Can be tested with the next release

## Summary by CodeRabbit

* **Chores**
  * Improved internal release automation to better detect normalized release-merge commits and select an appropriate baseline ref from multiple sources (env override, recent merge, release tag).
  * Updated test scripts to include tooling tests.

* **Tests**
  * Added comprehensive tests covering release-commit detection, baseline selection precedence, recent-merge discovery, and related edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/cowprotocol/cowswap/pull/7556?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tools/scripts/commits-to-changesets.test.mjs [hunk 1]
```diff
@@ -186,3 +188,89 @@ feat: bring cow.fi back (sdk) (#7359)
     assert.ok(out.startsWith('---\n"@scope/name": patch\n---\n'))
   })
 })
+
+describe('isReleaseCommitSubject', () => {
```
Comment (limitofzero) [inline tools/scripts/commits-to-changesets.test.mjs:178]: nitpick: will be good to add test case also for findRecentReleasePrCommit

File: tools/scripts/commits-to-changesets.test.mjs [hunk 2]
```diff
@@ -6,6 +6,8 @@ import {
   resolveAffectedPackages,
   changesetFilename,
   changesetContent,
+  isReleaseCommitSubject,
+  resolveBaselineRef,
 } from './commits-to-changesets-lib.mjs'
 
 describe('parseConventionalCommit', () => {
```
Comment (limitofzero) [inline tools/scripts/commits-to-changesets.test.mjs:14]: These tests aren't run at all, would be great to add something like: 
`"test:scripts": "node --test tools/scripts/*`
in package.json and check them in CI/locally

**Review by limitofzero — APPROVED** [review]

File: tools/scripts/commits-to-changesets.test.mjs [hunk 3]
```diff
@@ -186,3 +188,89 @@ feat: bring cow.fi back (sdk) (#7359)
...
+  it('env override wins over everything else', () => {
+    const r = resolveBaselineRef({
+      envBaseline: 'abc123',
+      latestReleaseTag: 'release-20260519T121234Z',
+      releasePrCommit: 'def456',
+    })
+    assert.deepEqual(r, { ref: 'abc123', source: 'env' })
+  })
+
+  it('release-PR merge commit wins over the release tag (the race-condition fix)', () => {
```
Comment (Danziger) [inline tools/scripts/commits-to-changesets.test.mjs]: Nitpick: I guess this "(the race-condition fix)" was added by AI based on the conversation with it. Not sure it makes sense to have it there.

**Review by Danziger — APPROVED** [review]

File: tools/scripts/commits-to-changesets.test.mjs (hunk 1 above)
Comment (shoom3301) [reply tools/scripts/commits-to-changesets.test.mjs:178]: Thanks! Added tests

File: tools/scripts/commits-to-changesets.test.mjs (hunk 2 above)
Comment (shoom3301) [reply tools/scripts/commits-to-changesets.test.mjs:14]: Thanks! Added it

File: tools/scripts/commits-to-changesets.test.mjs (hunk 3 above)
Comment (shoom3301) [reply tools/scripts/commits-to-changesets.test.mjs]: Thanks! Agree, it shouldn't be there (too specific).
Deleted it.
